### DISCUSSION TITLE
unable to upload file on system

### DISCUSSION THREAD

**Discussion by iprime2** [vercel/next.js#63175] [Help]
### Summary

I am trying to upload file like pdf, image, docx but it throws error, the code and error as follows.

```
import formidable from "formidable";
import path from "path";
import fs from "fs/promises";
import { NextResponse } from "next/server";
import { NextApiRequest } from "next";

export const config = {
  api: {
    bodyParser: false,
  },
};

const readFile = (
  req: NextApiRequest,
  saveLocally?: boolean
): Promise<{ fields: formidable.Fields; files: formidable.Files }> => {
  const options: formidable.Options = {};
  if (saveLocally) {
    options.uploadDir = path.join(process.cwd(), "public/documents");
    options.filename = (name, ext, path, form) => {
      return Date.now().toString() + "_" + path.originalFilename;
    };
  }
  const form = formidable(options);

  return new Promise((resolve, reject) => {
    form.parse(req, (err, fields, files) => {
      if (err) reject(err);
      resolve({ fields, files });
    });
  });
};

export async function POST(req: NextApiRequest, res: Response) {
  try {
    console.log(req);
    try {
      await fs.readdir(path.join(process.cwd() + "/public" + "/documents"));
    } catch (error) {
      await fs.mkdir(path.join(process.cwd() + "/public" + "/documents"));
    }

    await readFile(req, true);

    return NextResponse.json({ Message: "success", status: 200 });
  } catch (error) {
    console.log("[FILE_UPLOAD_ERROR]");
    console.error("Error:", error);
    return NextResponse.json({ Message: "Failed", status: 500 });
  }
}
```

frontend
![image](https://github.com/vercel/next.js/assets/29702609/58ab86cd-6734-4f9c-b241-edbde903fd19)

Error:
![image](https://github.com/vercel/next.js/assets/29702609/d869b243-dd90-44ce-8d7b-183da5357334)

**Comment by icyJoseph** [comment]
Hi,

There's a couple of things here. 

In App Router, the Route handlers, have a different signature than Page Route API Handlers.

So the body parser disabling, is not a thing anymore.

The Route handler is also called with just one parameter, and that is a Request, literally Web API Request: https://developer.mozilla.org/en-US/docs/Web/API/Request, which is extended by NextRequest, to gain more developer friendly APIs.

Moreover, in Route Handlers, you can just do `const formData = await req.formData()`, https://developer.mozilla.org/en-US/docs/Web/API/Request/formData

And that'll have the form fields filled in, so you don't need that many supporting libraries anymore.
